File: Hash-Based Data Structures/HashMaps.py
# Python
class Node(object):
    def __init__(self, next=None, key=None, value=None):
        self.next = next
        self.key = key
        self.value = value

class LinkedListMap(object):

    def __init__(self, head = None):
        self.head = head

    def prepend(self, key, value):
        if (self.head == None):
            self.head = Node(None, key, value)
            return
        tempHead = self.head
        self.head = Node(None, key, value)
        self.head.next = tempHead

    def delete(self, key):
        if (self.head == None):
            return
        if (self.head.key == key):
            self.head = self.head.next
            return
        current = self.head
        while (current != None and current.next != key):
            current = current.next
        current.next = current.next.next

    def findValue(self, key):
        current = self.head
        while (current != None and current.key != key):
            current = current.next
        if (current == None):
            return -1
        return current.value

    def print(self):
        current = self.head
        if (current == None):
            return
        print("{", end = "")
        while(current != None):
            print("[" + current.key + " : " + str(current.value) + "]", end="")
            current = current.next
        print("}")

    def getHead(self):
        if (self.head == None):
            return None
        return self.head

class HashMap(object):
    def __init__(self, size = 0):
        self.hashmap = list()
        self.hashmap = [None] * size;
        self.size = size

    def hash(self, str):
        hash = 0
        leftShiftValue = 1

        for x in range(0, 2):
            if (len(str) > x):
                hash = ord(str[x]) + ((hash << leftShiftValue) - hash)
        return hash

    def add(self, key, value):
        index = self.hash(key)
        if (self.hashmap[index] == None):
            self.hashmap[index] = LinkedListMap()
        if (self.hashmap[index].findValue(key) != -1):
            return
        self.hashmap[index].prepend(key, value)

    def delete(self, key):
        index = self.hash(key)
        if (self.hashmap[index] == None):
            return
        self.hashmap[index].delete(key)

    def get(self, key):
        index = self.hash(key)
        return self.hashmap[index].findValue(key)

    def print(self):
        for x in range(0, self.size):
            if (self.hashmap[x] != None):
                self.hashmap[x].print()


h = HashMap(253)
h.add("Hlavac", 735)
h.add("Strong", 145)
h.add("Dao", 984)
h.add("Dao", 118)
h.add("Dang", 567)
print(h.get("Dao"))
h.print()
h.delete("Strong")
h.delete("Strong")
h.print()
h.add("A", 123)
h.add("", 456)
h.add("%", 789)
h.add("~~", 5)
h.print()

File: Hash-Based Data Structures/HashMaps.cpp
//Node.hpp
struct Node {
    public:
        int value;
        std::string key;
        Node* next;
        Node(std::string k, int val);
};

//Node.cpp
#include <iostream>
#include "Node.hpp"
using namespace std;
Node::Node(std::string k, int val){
    key = k;
    value = val;
    next = NULL;
}

//LinkedListMap.hpp
#include "Node.cpp"
#include <iostream>
#include <vector>
using namespace std;
class LinkedListMap{
    public:
        Node* head;
        LinkedListMap();
        void prepend(std::string key, int value);
        void deleteN(std::string key);
        int findValue(std::string key);
        void print();
        Node* getHead();
};

//LinkedListMap.cpp
#include <iostream>
#include "LinkedListMap.hpp"
using namespace std;

LinkedListMap::LinkedListMap(){
    head = NULL;
}

void LinkedListMap::prepend(std::string key, int value){
    if (head == NULL){
        head = new Node(key, value);
        return;
    }
    Node* tempHead = head;
    head = new Node(key, value);
    head->next = tempHead;
}

void LinkedListMap::deleteN(std::string key){
    if (head == NULL){
        return;
    }
    if (head->key == key){
        head = head->next;
        return;
    }
    Node* current = head;
    while (current != NULL && (current->next->key != key)){
        current = current->next;
    }
    current->next = current->next->next;
}

int LinkedListMap::findValue(std::string key){
    Node* current = head;
    while (current != NULL && (current->key != key)){
        current = current->next;
    }
    if (current == NULL){
        return -1;
    }
    return current->value;
}

void LinkedListMap::print(){
    Node* current = head;
    if (current == NULL){
        return;
    }
    std::cout << "{";
    while (current != NULL){
        int i = current->value;
        std::string s = std::to_string(i);
        std::cout << "[" << current->key << " : " << s << "]";
        current = current->next;
    }
    std::cout << "} \n";
}

Node* LinkedListMap::getHead(){
    if (head == NULL){
        return NULL;
    }
    return head;
}

//HashMap.hpp
#include "LinkedListMap.cpp"
using namespace std;
class HashMap{
    public:
        std::vector<LinkedListMap> hashmap;
        HashMap(int size);
        int hash(std::string str);
        void add(std::string key, int value);
        void deleteN(std::string key);
        int get(std::string key);
        void print();
};

//HashMap.cpp
#include "HashMap.hpp"
#include <iostream>
using namespace std;
HashMap::HashMap(int size){
    std::vector<LinkedListMap> s(size);
    hashmap = s;
}

int HashMap::hash(std::string str){
    int hash = 0;
    int leftShiftValue = 1;

    for (int i = 0; i < 2; i++){
        if (str.length() > i){
            hash = int(str.at(i)) + ((hash << leftShiftValue) - hash);
        }
    }
    return hash;
}

void HashMap::add(std::string key, int value){
    int index = hash(key);
    if (hashmap[index].head == NULL){
        LinkedListMap l;
        hashmap[index] = l;
    }
    if (hashmap[index].findValue(key) != -1){
        return;
    }
    hashmap[index].prepend(key, value);

}

void HashMap::deleteN(std::string key){
    int index = hash(key);
    if (hashmap[index].head == NULL){
        return;
    }
    hashmap[index].deleteN(key);
}

int HashMap::get(std::string key){
    int index = hash(key);
    return hashmap[index].findValue(key);
}

void HashMap::print(){
    for (int i = 0; i < hashmap.size(); i++){
        if (hashmap[i].head != NULL){
            hashmap[i].print();
        }
    }
}

//main.cpp
#include <iostream>
#include "HashMap.cpp"
using namespace std;

void printList(){
    std::vector<int> ll;
    for(int i = 0; i < ll.size(); i++){
        std::cout << ll[i] << " -> ";
    }
    std::cout << "\n";

}

int main(){
    HashMap h = HashMap(253);
    h.add("Hlavac", 735);
		h.add("Strong", 145);
		h.add("Dao", 984);
		h.add("Dao", 118);
		h.add("Dang", 567);
		std::cout << h.get("Dao") << "\n";
		h.print();
		h.deleteN("Strong");
		h.deleteN("Strong");
		h.print();
		h.add("A", 123);
		h.add("", 456);
		h.add("%", 789);
		h.add("~~", 5);
		h.print();
}
